File: tests/golden/expected/03-generics.go
package main

import "fmt"

func Identity[T any](arg T) T {
	return arg
}

func Map[T any, U any](arr []T, fn func(T) U) []U {
	var result = []U{}
	for _, item := range arr {
		result = append(result, fn(item))
	}
	return result
}

type Container[T any] interface {
	Value() T
	GetValue() T
	SetValue(value T)
}

type Box[T any] struct {
	Value T
}

func NewBox[T any](value T) *Box[T] {
	return &Box[T]{
		Value:     value,
	}
}




func (b *Box[T]) GetValue() T {
	return b.Value
}

func (b *Box[T]) SetValue(value T) {
}

func MapBox[T any, U any](b *Box[T], fn func(T) U) *Box[U] {
	return NewBox(fn(b.Value))
}

type Lengthwise interface {
	Len() int
}

func LogLength[T Lengthwise](arg T) T {
	fmt.Println(arg.Len())
	return arg
}

type Pair[K any, V any] struct {
	Key       K
	Value     V
}

func NewPair[K any, V any](key K, value V) *Pair[K, V] {
	return &Pair[K, V]{
		Key:       key,
		Value:     value,
	}
}




func (p *Pair[K, V]) GetKey() K {
	return p.Key
}

func (p *Pair[K, V]) GetValue() V {
	return p.Value
}

type Response[T any] struct {
	Data      T
	Status    int
	Message   string
}

type IsString[T any] interface{}

type IsArray[T any] interface{}

var num = Identity(42)
var str = Identity("hello")
var arr = Map([]int{1, 2, 3}, func(x int) int { return x * 2 })
var box = NewBox(10)
var pair = NewPair("key", "value")

File: tests/golden/03-generics.ts
/**
 * 測試 03: 泛型
 */

// 泛型函式
function identity<T>(arg: T): T {
  return arg;
}

function map<T, U>(arr: T[], fn: (item: T) => U): U[] {
  const result: U[] = [];
  for (const item of arr) {
    result.push(fn(item));
  }
  return result;
}

// 泛型介面
interface Container<T> {
  value: T;
  getValue(): T;
  setValue(value: T): void;
}

// 泛型類別
class Box<T> implements Container<T> {
  constructor(public value: T) {}

  getValue(): T {
    return this.value;
  }

  setValue(value: T): void {
    this.value = value;
  }

  map<U>(fn: (value: T) => U): Box<U> {
    return new Box(fn(this.value));
  }
}

// 泛型約束
interface Lengthwise {
  length: number;
}

function logLength<T extends Lengthwise>(arg: T): T {
  console.log(arg.length);
  return arg;
}

// 多個型別參數
class Pair<K, V> {
  constructor(
    public key: K,
    public value: V
  ) {}

  getKey(): K {
    return this.key;
  }

  getValue(): V {
    return this.value;
  }
}

// 泛型預設值
interface Response<T = any> {
  data: T;
  status: number;
  message: string;
}

// 條件型別 (簡化版)
type IsString<T> = T extends string ? true : false;
type IsArray<T> = T extends any[] ? true : false;

// 使用範例
const num = identity(42);
const str = identity("hello");
const arr = map([1, 2, 3], x => x * 2);
const box = new Box<number>(10);
const pair = new Pair("key", "value");

export { identity, map, Box, Pair, logLength };